fastembed/splitter.py

```diff
@@ -0,0 +1,206 @@
+        chunk_size: int = 4000,
+        chunk_overlap: int = 200,
+        strip_whitespace: bool = True,
+        separators: Optional[List[str]] = None,
+        keep_separator: bool = True,
+        is_separator_regex: bool = False,
+    ) -> None:
+        """Create a new TextSplitter."""
+
+        if not isinstance(tokenizer, Tokenizer):
+            raise ValueError("Tokenizer received was not an instance of tokenizers.Tokenizer")
+
+        def _tokenizer_length(text: str) -> int:
+            return len(tokenizer.encode(text))
+
+        super().__init__(
+            chunk_size=chunk_size,
+            chunk_overlap=chunk_overlap,
+            length_function=_tokenizer_length,
+            strip_whitespace=strip_whitespace,
+            keep_separator=keep_separator,
+        )
+        self._separators = separators or ["\n\n", "\n", " ", ""]
+        self._is_separator_regex = is_separator_regex
+
+    def _split_text(
+        self, text: str, separators: List[str], chunk_size: Optional[int] = None, chunk_overlap: Optional[int] = None
+    ) -> List[str]:
+        """Split incoming text and return chunks."""
+
+        chunk_size = chunk_size or self._chunk_size
+        chunk_overlap = chunk_overlap or self._chunk_overlap
+
+        final_chunks = []
+        # Get appropriate separator to use
+        separator = separators[-1]
+        new_separators = []
+        for i, _s in enumerate(separators):
+            _separator = _s if self._is_separator_regex else re.escape(_s)
+            if _s == "":
+                separator = _s
+                break
+            if re.search(_separator, text):
+                separator = _s
+                new_separators = separators[i + 1 :]
+                break
+
+        _separator = separator if self._is_separator_regex else re.escape(separator)
+        splits = _split_text_with_regex(text, _separator, self._keep_separator)
+
+        # Now go merging things, recursively splitting longer texts.
+        _good_splits = []
+        _separator = "" if self._keep_separator else separator
+        for s in splits:
+            if self._length_function(s) < chunk_size:
+                _good_splits.append(s)
+            else:
+                if _good_splits:
+                    merged_text = self._merge_splits(_good_splits, _separator, chunk_size, chunk_overlap)
+                    final_chunks.extend(merged_text)
+                    _good_splits = []
+                if not new_separators:
+                    final_chunks.append(s)
+                else:
+                    other_info = self._split_text(s, new_separators)
+                    final_chunks.extend(other_info)
+        if _good_splits:
+            merged_text = self._merge_splits(_good_splits, _separator)
+            final_chunks.extend(merged_text)
+        return final_chunks
+
+    def split_text(self, text: str, chunk_size: Optional[int] = None, chunk_overlap: Optional[int] = None) -> List[str]:
```

Comment:
I'm not quite sure why we need this wrapper function? Let's just rename the _split_text to split_text?